plugin.py

```diff
@@ -0,0 +1,156 @@
+import sublime
+
+import os
+import re
+import base64
+from urllib.parse import urlparse
+import weakref
+
+from LSP.plugin.core.protocol import Request, Location
+from LSP.plugin.core.sessions import Session
+from LSP.plugin.core.typing import Any, Callable, Dict, List, Mapping, Optional, Tuple, Union
+from LSP.plugin.core.registry import LspTextCommand
+from LSP.plugin import register_plugin, unregister_plugin
+from lsp_utils import NpmClientHandler
+
+
+class LspGrammarlyCommand(LspTextCommand):
+    session_name = 'grammarly'
+    msg_prefix = "LSP-Grammarly: "
+
+    def error_message(self, msg: str) -> None:
+        sublime.error_message(self.msg_prefix + msg)
+
+    def message_dialog(self, msg: str) -> None:
+        sublime.message_dialog(self.msg_prefix + msg)
+
+
+class LspGrammarlyExecuteLoginCommand(LspGrammarlyCommand):
+    cmd = "$/getOAuthUrl"
+    cmd_callback = "$/handleOAuthCallbackUri"
+
+    internal_redirect_uri = "vscode://znck.grammarly/auth/callback"
+    external_redirect_uri = internal_redirect_uri
+    redirect_uri = internal_redirect_uri
+    weaksession = None  # type: Optional[weakref.ref[Session]]
+
+    def run(self, edit: sublime.Edit) -> None:
+        def run_async() -> None:
+            session = self.session_by_name()
+            if not session:
+                return
+            self.weaksession = weakref.ref(session)
+            request = Request(self.cmd, self.redirect_uri, None, progress=True)
+            session.send_request(request, self._handle_response_async,
+                                 lambda _: self.error_message("Failed to start login process"))
+        sublime.set_timeout_async(run_async)
+
+    def _handle_response_async(self, response: Union[List[Location], None]) -> None:
+        link = re.sub("state=[^&]*",
+                      "state=" + base64.b64encode(bytes(self.external_redirect_uri, 'utf-8')).decode('ascii'),
+                      str(response))
+        session = self.weaksession and self.weaksession()
+        if session:
+            session.window.show_input_panel(
+                "Open this url in the browser and paste in its place the redirection:", link,
+                lambda r: self._send_response_async(r), None, lambda: self._send_response_async(None))
```

Comment:
Actually I've just tried this command and the flow seems weird/confusion/non-working...

After authorizing Grammarly, a new empty tab opens that wants to open a native application. When I click cancel on that dialog, I'm left with no URL anywhere that I could copy back into ST.

I'm using Firefox, I wonder if Chrome would still show the URL in the URL field...

Reply:
And also the second time this command is triggered, it immediately triggers `on_change` with the initial value and triggers an error.
The first time it's invoked it also triggers `on_change` but with value `None`.

Reply:
Tried in Chrome now and it indeed works there.

Could we maybe just open the URL automatically using python `webbrowser` and only ask to insert the response URL in the field?

Reply:
i added the webbrowser open + an alternative that uses the VSCode package's redirect-link to do the trick without needing anyone to copy paste the URL. The drawback of the latter is that it can't directly redirect to localhost and instead it has to go through a netlify function of the VSCode plugin, so I left both and a note the latter goes through a third party.